celery/backends/asynchronous.py

```diff
@@ -78,23 +83,37 @@ def _send_drain_complete_event(self):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        self._mutex = threading.Lock()
         self._started = threading.Event()
         self._stopped = threading.Event()
         self._shutdown = threading.Event()
         self._create_drain_complete_event()
 
     def run(self):
         self._started.set()
-        while not self._stopped.is_set():
-            try:
-                self.result_consumer.drain_events(timeout=1)
-                self._send_drain_complete_event()
-                self._create_drain_complete_event()
-            except socket.timeout:
-                pass
-        self._shutdown.set()
+
+        try:
+            while not self._stopped.is_set():
+                try:
+                    self.result_consumer.drain_events(timeout=1)
+                    self._send_drain_complete_event()
+                    self._create_drain_complete_event()
+                except socket.timeout:
+                    pass
+        except Exception as e:
+            with self._mutex:
+                self._exc = e
+        finally:
+            self._shutdown.set()
 
     def start(self):
+        if self._shutdown.is_set():
+            with self._mutex:
+                if self._exc is not None:
+                    raise self._exc
+                else:
+                    raise Exception(E_CELERY_RESTART_REQUIRED)
```

Comment:
Also an open question here: do we want to tell clients that they need to restart the celery app, or would we rather attempt to restart the greenlet run loop within celery if the greenlet is forced to exit from unhandled exceptions?